deployment: add helm charts, configuration and openshift docs (#30)
inveniosoftware/docs-invenio-rdm: docs/deployment/configuration.md

# Configuration

> This section explains the current configuration options that are available for the different components deployed by these Helm charts.

Reviewer: See my previous comments about what "these Helm charts" refer to?

# Configuration

> This section explain the current configuration options that are available for the different components that are deployed but the Helm charts.

Reviewer: ```suggestion
This section explains the configuration options that are available for the different components deployed by the InvenioRDM Helm chart.
```

(it's one chart for InvenioRDM, right?)

Author: not really, is just the image that is configurable. Once we set it up we will put it in the docs, but this charts suit all Invenio instances.

Reviewer: To be clear, in the updated configuration.md, what do "these Helm charts" refer to? Is the invenio chart an amalgamation of charts (or maybe what I am saying makes no sense in Kubernetes world?)

## Global

> The is only one mandatory configuration, which is the host name, for example:

Reviewer: ```suggestion
There is only one mandatory configuration: the host name.
```

> The is only one mandatory configuration, which is the host name, for example:

Author: Nope, it is in a block of code just below :)

```yaml
host: your-rdm-instance.com
```

> Moreover, the services can be deployed along, note that it is recommended to deploy separatelly Elasticsearch and PostgreSQL for a production deployment.

Reviewer: ```suggestion
Moreover, the services can be deployed along. Note that it is recommended to deploy Elasticsearch and PostgreSQL separately for a production deployment.
```

Therefore, by default only `redis` and `rabbitmq` are enabled. Example configuration:

``` yaml
postgresql:
  inside_cluster: false

elasticsearch:
  inside_cluster: false
```

!!! info "inside_cluster availability
    Note that the `inside_cluster` variable is supported for `redis`, `rabbitmq`, `elasticsearch`, `postgresql` and `haproxy`. The rest of the components
    are mandatory.

## HAProxy

> You can change the number of connections allowed by the haproxy with the `maxconn` variable:

Reviewer: ```suggestion
You can change the number of connections allowed by haproxy with the `maxconn` variable:
```

```
haproxy:
  maxconn: 100
```

!!! warning "Only one parent element"
    By default the HAProxy is enabled, `inside_cluster` has the value `true`, nonetheless if you decide to set it, you only need
    to specify once the `haproxy` parent in the yaml file. Otherwise the last one will override the previous. It should be something like:
    ``` yaml
    haproxy:
        inside_cluster: true
        maxconn: 100

>     ```

Reviewer: I don't think I understand this part. Could you rephrase it?  :smiley_cat: (haproxy is enabled by default... so why do I need to specify it? What do I "set" and what is the "last one" that "will override the previous"?)

Author: It is a case that should not happen. If it is enabled you only, might, need to change the maxcoon. Otherwise just disable it. My point was more of a yaml syntax point. That it should be:

```yaml
haproxy:
    inside_cluster: true
    maxconn: 100
```

because if you do:

```yaml
haproxy:
    inside_cluster: true

haproxy:
    maxconn: 100
```

only the second one applies, but in this case doesn't really make sense since it is active by default. So I removed the paragraph.

Reviewer: Ah got it.

## Nginx

> The charts allow you to configure the amount of connections per nginx node (replica) and the amount of nodes:

Reviewer: 1 chart <==> 1 package . "chart" is singular or is it "charts" because multiple chart are combined together?

Author: I have seen both in the literature, I have not seen a naming convention. Will look for it.

```
nginx:
  max_conns: 100
  replicas: 2
```

## Web nodes

> The web nodes host the WSGI application, in order to be scalable you can configure the number of "nodes", called replicas, how many processes do each node run and with how many threads per process. The only mandatory parameter is the docker image (`image`) that should get as value the url where to pull the image from.

Reviewer: ```suggestion
The web nodes host the WSGI application. In order to be scalable, you can configure the number of "nodes", called replicas, with how many processes each node runs and with how many threads per process. The only mandatory parameter is the docker image (`image`) that should get as value the url where to pull the image from.
```

> In addition, you can add automatic scaling, by setting minimum and maximum replicas and the threshold of cpu usage in which a new node should be spawned. For example, with a threshold of 65%, it meand that when the average CPU utilization of the nodes reaches 65% a new node will de spawned, till it reaches the setted maximum:

Reviewer: ```suggestion
In addition, you can add automatic scaling by setting minimum (`min_web_replicas`) and maximum (`max_web_replicas`) replicas, and the cpu usage threshold in percentage (`scaler_cpu_utilization`) that spawns a new node. For example, with a `scaler_cpu_utilization` value of 65, it means that when the average CPU utilization of the nodes reaches 65%, a new node will be spawned. This process will repeat itself until the maximum number of replicas has been reached:
```

``` yaml
web:
  image: your/invenio-image
  replicas: 6
  uwsgi:
    processes: 6
    threads: 4
  autoscaler:
    enabled: false
    # Scale when CPU usage gets to
    scaler_cpu_utilization: 65
    max_web_replicas: 10
    min_web_replicas: 2
```

## Worker nodes

> Finally, the worker nodes. By default they are enabled, but you can cancel their deployment by setting `enabled` to `false`. If enabled, in the same fashion than the web nodes

Reviewer: ```suggestion
Finally, the worker nodes. By default they are enabled, but you can cancel their deployment by setting `enabled` to `false`. If enabled, they require an `image` like the web nodes.
```

> they require an `image`.

Reviewer: ```suggestion
```

> In addition, you can configure how many worker nodes (replicas) will be deployed, which the application they will run, with which concurrency level and the logging level.

Reviewer: ```suggestion
In addition, you can configure the number of worker nodes (replicas) to be deployed, the application they will run, their concurrency level and their logging level.
```
